pyro/infer/mcmc/hmc.py

```diff
@@ -0,0 +1,197 @@
+import random
+
+import torch
+from torch.autograd import Variable
+
+import pyro
+import pyro.distributions as dist
+import pyro.poutine as poutine
+from pyro.distributions.util import torch_zeros_like, torch_ones_like
+from pyro.infer import TracePosterior
+from pyro.poutine import Trace
+from tests.common import assert_equal
+
+
+def verlet_integrator(z, r, grad_potential, step_size, num_steps):
+    """
+    Velocity Verlet integrator.
+
+    :param z: trace object containing current values for the sample sites
+    :param r: dictionary of sample site names and corresponding momenta
+    :param grad_potential: function that returns gradient of the potential given z
+        for each sample site
+    :return: (z_next, r_next) having same types as (z, r)
+    """
+    # deep copy the current state - (z, r)
+    z_next = clone_trace(z)
+    r_next = {k: v.clone() for k, v in r.items()}
+    grads = grad_potential(z_next)
+    for _ in range(num_steps):
+        for site_name, node in z_next.iter_stochastic_nodes():
+            # r(n+1/2)
+            r_next[site_name] = r_next[site_name] + 0.5 * step_size * (-grads[site_name])
+            # z(n+1)
+            node['value'] = node['value'] + step_size * r_next[site_name]
```

Comment:
I'm surprised you don't `.detach()` here. Wouldn't this implementation propagate gradients far far back in time, through all previous time steps? If so, this might be a possible cause of segfault on PyTorch 0.2.

Reply:
Oops..that definitely seems problematic. Let me take a look. Thanks for pointing out!